core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServices.java

```diff
@@ -108,9 +111,52 @@
      * Returns the {@link PathMapping} instance mapped to {@code method}.
      */
     private static PathMapping pathMapping(String pathPrefix, Method method) {
-        Path mapping = method.getAnnotation(Path.class);
-        String mappedTo = ArmeriaHttpUtil.concatPaths(pathPrefix, mapping.value());
-        return PathMapping.of(mappedTo);
+        pathPrefix = ensureAbsolutePath(pathPrefix, "pathPrefix");
+        if (!pathPrefix.endsWith("/")) {
+            pathPrefix += '/';
+        }
+
+        final String pattern = method.getAnnotation(Path.class).value();
+        if (pattern == null) {
+            throw new NullPointerException("The value of @Path must be non-null.");
+        }
+
+        final PathMapping mapping = PathMapping.of(pattern);
+        if ("/".equals(pathPrefix)) {
+            // pathPrefix is not specified or "/".
+            return mapping;
+        }
+
+        if (pattern.startsWith("exact:")) {
+            // 6 = "exact:".length()
+            return PathMapping.ofExact(concatPaths(pathPrefix, pattern.substring(6)));
+        }
+
+        if (pattern.startsWith("prefix:")) {
+            // 7 = "prefix:".length()
+            return PathMapping.ofPrefix(concatPaths(pathPrefix, pattern.substring(7)));
+        }
+
+        if (pattern.startsWith("glob:")) {
+            final String glob = pattern.substring(5); // 5 = "glob:".length()
+            if (glob.startsWith("/")) {
+                return PathMapping.ofGlob(concatPaths(pathPrefix, glob));
+            } else {
+                return PathMapping.ofGlob(concatPaths(pathPrefix, "/**/" + glob));
+            }
+        }
+
+        if (pattern.startsWith("regex:")) {
+            return new PrefixRegexPathMapping(pathPrefix, (RegexPathMapping) mapping);
+        }
+
+        if (pattern.startsWith("/")) {
+            // Default pattern
+            return PathMapping.of(concatPaths(pathPrefix, pattern));
+        }
+
+        // Should never reach here because we validated the path pattern.
+        throw new Error();
```

Comment:
I think it's better throwing an `Error` here because we checked the precondition with `PathMapping.of()` above and thus it is impossible to reach here.